### ISSUE TITLE
Could we monitoring the audio after capturing?

### ISSUE THREAD

**Issue by huangjinshe** [libsdl-org/SDL#8541]
I'm trying to write a sample code using SDL3, which **output the audio after capture microphone** (output the audio which captured by microphone).

So I start:
```c
int main(int argc, char* argv[]) {
    if (SDL_Init(SDL_INIT_AUDIO) < 0) {
        SDL_Log("SDL could not initialize! SDL_Error: %s\n", SDL_GetError());
        return -1;
    }


    SDL_AudioSpec desired_spec;

    SDL_zero(desired_spec);
    desired_spec.freq = 44800; 
    desired_spec.format = SDL_AUDIO_F32; 
    desired_spec.channels = 2; 


    // Output
    SDL_AudioStream* stream_output = SDL_CreateAudioStream(&desired_spec, NULL);
    SDL_AudioDeviceID output_audio_device = SDL_OpenAudioDevice(SDL_AUDIO_DEVICE_DEFAULT_OUTPUT, &desired_spec);
    int result = SDL_BindAudioStream(output_audio_device, stream_output);


    // Input (from microhpone)
    SDL_AudioStream* stream_input = SDL_CreateAudioStream(&desired_spec, NULL);
    SDL_AudioDeviceID input_audio_device = SDL_OpenAudioDevice(SDL_AUDIO_DEVICE_DEFAULT_CAPTURE, &desired_spec);
    result = SDL_BindAudioStream(output_audio_device, stream_input);

    SDL_SetAudioStreamGetCallback(stream_input, capture_audio_callback, stream_output);

    // Main loop
    bool quit = false;

    SDL_Event e;
    while (!quit) {
        while (SDL_PollEvent(&e) != 0) {
            if (e.type == SDL_EVENT_QUIT) {
                quit = true;
            }
        }
    }
    SDL_Quit();
    return 0;
}
```
Bacically I use new audio subsystem style create 2 audio streams, 1 for capture, 1 for output.
And then I trying to use callback capture the audio and output it to the output stream:

```c
void capture_audio_callback(void* userdata, SDL_AudioStream* input_stream, int additional_amount, int total_amount) {

    SDL_AudioStream* output_stream = (SDL_AudioStream*)userdata;

    void* buf = (Uint8*)SDL_calloc(1, total_amount);
    int res = SDL_GetAudioStreamData(input_stream, &buf, total_amount);

    SDL_PutAudioStreamData(output_stream, buf, total_amount);
    SDL_free(&buf);
}
```

But it didn't work,  did I misunderstanding the new audio subsystem?

Why using like this?
I think a lot of scenario we might need to real time outputting the audio for ourselves,  like using Live stream app, we need to hear what we saying, and what people listening.   Sometimes we need to pre process our audio data, add some echo, or some effects, and then output it to people, but we also need to output it to ourselves for knowing the result before.(So could we process the data in the callback before output?  that's my second question).

**Comment by slouken** [comment]
Your callback should be like this:
```c
void capture_audio_callback(void* userdata, SDL_AudioStream* input_stream, int additional_amount, int total_amount) {

    SDL_AudioStream* output_stream = (SDL_AudioStream*)userdata;

    void* buf = SDL_malloc(total_amount);
    int res = SDL_GetAudioStreamData(input_stream, buf, total_amount);

    SDL_PutAudioStreamData(output_stream, buf, res);
    SDL_free(buf);
}
```

**Comment by huangjinshe** [comment]
Unfortunately, it would crash with error: `0xC00000FD: Stack overflow`

**Comment by slouken** [comment]
What's the call stack?

**Comment by icculus** [comment]
>  SDL_SetAudioStreamGetCallback

That should be  SDL_SetAudioStreamPutCallback. The callback fires when data is _put_ into the stream, which is what SDL is doing when new data arrives from the hardware.

**Comment by icculus** [comment]
> What's the call stack?

An infinite recursion in his callback, because he was getting from a get callback.  :)

**Comment by huangjinshe** [comment]
@icculus  If change it to `SDL_SetAudioStreamPutCallback(stream_input, capture_audio_callback, stream_output);` nothing would happen.  Are you mean I need to create 2 callback for input stream and output stream? `SDL_SetAudioStreamPutCallback` for output steram. `SDL_SetAudioStreamGetCallback` for input stream?

**Comment by icculus** [comment]
Okay, I built the original program, and there are a few issues that are easily fixed:

First, the input stream was bound to the output device, through a copy/paste error:

This line...

```c
    result = SDL_BindAudioStream(output_audio_device, stream_input);
```

...should use `input_audio_device`

Next, the input device needs a destination format (binding it to an input device will set the source format):

This line...

```c
    SDL_AudioStream* stream_input = SDL_CreateAudioStream(&desired_spec, NULL);
```

Should be:

```c
    SDL_AudioStream* stream_input = SDL_CreateAudioStream(NULL, &desired_spec);
```

As I mentioned, the Get callback should be a Put callback:

```c
    SDL_SetAudioStreamGetCallback(stream_input, capture_audio_callback, stream_output);
```

...should be...

```c
    SDL_SetAudioStreamPutCallback(stream_input, capture_audio_callback, stream_output);
```


Finally, the callback is misusing malloc and free:

```c
    int res = SDL_GetAudioStreamData(input_stream, &buf, total_amount);
```

The `&buf` should be `buf` or this will write over the buf variable and whatever is on the stack after it, instead of writing to the allocated buffer.

```c
    SDL_free(&buf);
```

Same here. This should be `buf` and not `&buf`.


After these changes, I got sound; just be careful that you're using headphones, or this will cause feedback when the microphone picks up the sound that's playing.
